{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from skbio.stats.composition import clr\n",
    "import os\n",
    "import sys\n",
    "os.chdir('/home/alexis/UAM/picrust/DataSets/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Right</th>\n",
       "      <th>Left</th>\n",
       "      <th>EC-Number</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reaction</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DEOXYNUCLEOTIDE-3-PHOSPHATASE-RXN</th>\n",
       "      <td>Deoxy-Ribonucleosides Pi</td>\n",
       "      <td>Deoxy-Ribonucleoside-3P WATER</td>\n",
       "      <td>EC:3.1.3.34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RXN-20295</th>\n",
       "      <td>CPD-21889 CARBON-DIOXIDE WATER Oxidized-ferred...</td>\n",
       "      <td>CPD-21911 OXYGEN-MOLECULE Reduced-ferredoxins ...</td>\n",
       "      <td>EC:1.14.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RXN-11130</th>\n",
       "      <td>CPD-12101 CO-A</td>\n",
       "      <td>CPD-11571 ACETYL-COA</td>\n",
       "      <td>EC:2.3.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HYPONITRITE-REDUCTASE-RXN</th>\n",
       "      <td>PROTON CPD-1541 NADH</td>\n",
       "      <td>HYDROXYLAMINE NAD</td>\n",
       "      <td>EC:1.7.1.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RXN-19187</th>\n",
       "      <td>CPD-20683 ADP PROTON</td>\n",
       "      <td>CPD-20682 ATP</td>\n",
       "      <td>EC:2.7.1.M16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                               Right  \\\n",
       "Reaction                                                                               \n",
       "DEOXYNUCLEOTIDE-3-PHOSPHATASE-RXN                           Deoxy-Ribonucleosides Pi   \n",
       "RXN-20295                          CPD-21889 CARBON-DIOXIDE WATER Oxidized-ferred...   \n",
       "RXN-11130                                                             CPD-12101 CO-A   \n",
       "HYPONITRITE-REDUCTASE-RXN                                       PROTON CPD-1541 NADH   \n",
       "RXN-19187                                                       CPD-20683 ADP PROTON   \n",
       "\n",
       "                                                                                Left  \\\n",
       "Reaction                                                                               \n",
       "DEOXYNUCLEOTIDE-3-PHOSPHATASE-RXN                      Deoxy-Ribonucleoside-3P WATER   \n",
       "RXN-20295                          CPD-21911 OXYGEN-MOLECULE Reduced-ferredoxins ...   \n",
       "RXN-11130                                                       CPD-11571 ACETYL-COA   \n",
       "HYPONITRITE-REDUCTASE-RXN                                          HYDROXYLAMINE NAD   \n",
       "RXN-19187                                                              CPD-20682 ATP   \n",
       "\n",
       "                                      EC-Number  \n",
       "Reaction                                         \n",
       "DEOXYNUCLEOTIDE-3-PHOSPHATASE-RXN   EC:3.1.3.34  \n",
       "RXN-20295                            EC:1.14.15  \n",
       "RXN-11130                              EC:2.3.1  \n",
       "HYPONITRITE-REDUCTASE-RXN            EC:1.7.1.5  \n",
       "RXN-19187                          EC:2.7.1.M16  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Create Metabolic.net files for analysis in BiomeNet\n",
    "#Metabolic net files have the format reactant,reaxionID,product\n",
    "\n",
    "#Load reactions in the Metacyc Database, annotated with reactants and products\n",
    "EC_reactions = pd.read_csv('/home/alexis/UAM/picrust/default_files/metacyc_rxns_v24.csv', sep = '\\t', index_col = 'Reaction') \n",
    "EC_reactions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>function</th>\n",
       "      <th>SRR2071024</th>\n",
       "      <th>SRR2071075</th>\n",
       "      <th>SRR2029458</th>\n",
       "      <th>SRR2029459</th>\n",
       "      <th>SRR2029460</th>\n",
       "      <th>SRR7135742</th>\n",
       "      <th>SRR7135743</th>\n",
       "      <th>CIR_18</th>\n",
       "      <th>IL_18</th>\n",
       "      <th>...</th>\n",
       "      <th>PR9_18</th>\n",
       "      <th>ALC_19</th>\n",
       "      <th>CIR_19</th>\n",
       "      <th>ERR844085</th>\n",
       "      <th>ERR844086</th>\n",
       "      <th>ALC_20</th>\n",
       "      <th>My_20</th>\n",
       "      <th>Pty_20</th>\n",
       "      <th>V3_20</th>\n",
       "      <th>VT_20</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EC:1.1.1.1</td>\n",
       "      <td>569327.80</td>\n",
       "      <td>1358307.09</td>\n",
       "      <td>32853.85</td>\n",
       "      <td>31431.34</td>\n",
       "      <td>32776.49</td>\n",
       "      <td>108609.34</td>\n",
       "      <td>114356.08</td>\n",
       "      <td>2119.0</td>\n",
       "      <td>8145.25</td>\n",
       "      <td>...</td>\n",
       "      <td>7749.0</td>\n",
       "      <td>16234.75</td>\n",
       "      <td>2726.0</td>\n",
       "      <td>79297.66</td>\n",
       "      <td>232808.34</td>\n",
       "      <td>54481.17</td>\n",
       "      <td>45108.65</td>\n",
       "      <td>38240.35</td>\n",
       "      <td>24364.68</td>\n",
       "      <td>92869.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EC:1.1.1.100</td>\n",
       "      <td>1595044.88</td>\n",
       "      <td>4017067.99</td>\n",
       "      <td>75906.92</td>\n",
       "      <td>85869.36</td>\n",
       "      <td>91134.96</td>\n",
       "      <td>316157.74</td>\n",
       "      <td>319596.12</td>\n",
       "      <td>22631.5</td>\n",
       "      <td>27922.25</td>\n",
       "      <td>...</td>\n",
       "      <td>16719.0</td>\n",
       "      <td>30634.50</td>\n",
       "      <td>71197.0</td>\n",
       "      <td>178744.34</td>\n",
       "      <td>117803.02</td>\n",
       "      <td>106479.34</td>\n",
       "      <td>77172.98</td>\n",
       "      <td>144959.02</td>\n",
       "      <td>81021.79</td>\n",
       "      <td>138866.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EC:1.1.1.102</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>41.00</td>\n",
       "      <td>7.50</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EC:1.1.1.103</td>\n",
       "      <td>76164.33</td>\n",
       "      <td>222163.72</td>\n",
       "      <td>423.42</td>\n",
       "      <td>377.50</td>\n",
       "      <td>1257.00</td>\n",
       "      <td>20822.83</td>\n",
       "      <td>16663.33</td>\n",
       "      <td>176.0</td>\n",
       "      <td>1236.50</td>\n",
       "      <td>...</td>\n",
       "      <td>2846.0</td>\n",
       "      <td>230.75</td>\n",
       "      <td>181.0</td>\n",
       "      <td>2761.33</td>\n",
       "      <td>33.67</td>\n",
       "      <td>11111.00</td>\n",
       "      <td>2484.00</td>\n",
       "      <td>5552.00</td>\n",
       "      <td>7429.00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EC:1.1.1.105</td>\n",
       "      <td>120.00</td>\n",
       "      <td>790.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       function  SRR2071024  SRR2071075  SRR2029458  SRR2029459  SRR2029460  \\\n",
       "0    EC:1.1.1.1   569327.80  1358307.09    32853.85    31431.34    32776.49   \n",
       "1  EC:1.1.1.100  1595044.88  4017067.99    75906.92    85869.36    91134.96   \n",
       "2  EC:1.1.1.102        0.00        0.00        0.00       41.00        7.50   \n",
       "3  EC:1.1.1.103    76164.33   222163.72      423.42      377.50     1257.00   \n",
       "4  EC:1.1.1.105      120.00      790.00        0.00        0.00        0.00   \n",
       "\n",
       "   SRR7135742  SRR7135743   CIR_18     IL_18  ...   PR9_18    ALC_19   CIR_19  \\\n",
       "0   108609.34   114356.08   2119.0   8145.25  ...   7749.0  16234.75   2726.0   \n",
       "1   316157.74   319596.12  22631.5  27922.25  ...  16719.0  30634.50  71197.0   \n",
       "2        0.00        0.00      0.0      0.00  ...      0.0      0.00      0.0   \n",
       "3    20822.83    16663.33    176.0   1236.50  ...   2846.0    230.75    181.0   \n",
       "4        0.00        0.00      0.0      0.00  ...      0.0      0.00      0.0   \n",
       "\n",
       "   ERR844085  ERR844086     ALC_20     My_20     Pty_20     V3_20     VT_20  \n",
       "0   79297.66  232808.34   54481.17  45108.65   38240.35  24364.68   92869.4  \n",
       "1  178744.34  117803.02  106479.34  77172.98  144959.02  81021.79  138866.6  \n",
       "2       0.00       0.00       0.00      0.00       0.00      0.00       0.0  \n",
       "3    2761.33      33.67   11111.00   2484.00    5552.00   7429.00       0.0  \n",
       "4       0.00       0.00       0.00      0.00       0.00      0.00       0.0  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metagenome_unstrat = pd.read_csv('EC_metagenome_out/pred_metagenome_unstrat.tsv.gz', compression = 'gzip', sep = '\\t')\n",
    "metagenome_unstrat.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17\n",
      "1\n",
      "1\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "#Create the files for metabolic divergence analysis between samples\n",
    "#For each sample in the metagenome_file\n",
    "missing_ec = []\n",
    "for sample in metagenome_unstrat.iloc[:,1:].columns:\n",
    "    #Create a metabolic.net file\n",
    "    metabolic_net = open('/home/alexis/BiomeNet/BiomeNet/metabolic_nets/DataSets/'+sample+'.metabolic.net','w')\n",
    "    #Filter the reactions with non-zero abundance\n",
    "    nonzero_filter = metagenome_unstrat.loc[:, sample] > 0\n",
    "    temp_df = metagenome_unstrat.loc[nonzero_filter,['function', sample]]\n",
    "    temp_df.loc[:, sample] = clr(temp_df.loc[:, sample])\n",
    "    temp_df.loc[:, sample] = temp_df.loc[:, sample] - 5*temp_df.loc[:, sample].min()\n",
    "    #For each non-zero abundance reaction\n",
    "    n = 0\n",
    "    for index in temp_df.index:\n",
    "        #Filter all reactions in the metacyc database annotated with indexed EC-number in metagenome\n",
    "        EC_filter = EC_reactions.loc[:, 'EC-Number'] == temp_df.loc[index, 'function']\n",
    "        if not EC_reactions.loc[EC_filter].index.empty:\n",
    "            #For all reactions wich matched the EC-number\n",
    "            for rxn in EC_reactions.loc[EC_filter].index:\n",
    "                #repeat the process n times, where n is the abundace of the indexed reaction\n",
    "                for i in range(int(round(temp_df.loc[index, sample]))):\n",
    "                    #Map reactant,reaxionID,product\n",
    "                    for reactant in EC_reactions.loc[rxn, 'Right'].split():\n",
    "                        for product in EC_reactions.loc[rxn, 'Left'].split():\n",
    "                            metabolic_net.write(reactant+','+rxn+','+product+'\\n')\n",
    "                    metabolic_net.write('\\n')\n",
    "        else:\n",
    "            if temp_df.loc[index, 'function'] in missing_ec:\n",
    "                pass\n",
    "            else:\n",
    "                n += 1\n",
    "                missing_ec.append(temp_df.loc[index, 'function'])\n",
    "    print(n)\n",
    "    del temp_df\n",
    "    metabolic_net.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>function</th>\n",
       "      <th>taxon</th>\n",
       "      <th>taxon_abun</th>\n",
       "      <th>taxon_rel_abun</th>\n",
       "      <th>genome_function_count</th>\n",
       "      <th>taxon_function_abun</th>\n",
       "      <th>taxon_rel_function_abun</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ALC_19</td>\n",
       "      <td>EC:1.1.1.1</td>\n",
       "      <td>0746d4ed931f7a6c799e7b1e5f525cbc</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.642873</td>\n",
       "      <td>1</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.642873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ALC_19</td>\n",
       "      <td>EC:1.1.1.1</td>\n",
       "      <td>108039c87f18b4ab5b726b4417cefdca</td>\n",
       "      <td>473.0</td>\n",
       "      <td>10.485480</td>\n",
       "      <td>2</td>\n",
       "      <td>946.0</td>\n",
       "      <td>20.970960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ALC_19</td>\n",
       "      <td>EC:1.1.1.1</td>\n",
       "      <td>16d4ee39d4468351b3a62dbd64d29ed9</td>\n",
       "      <td>38.0</td>\n",
       "      <td>0.842385</td>\n",
       "      <td>1</td>\n",
       "      <td>38.0</td>\n",
       "      <td>0.842385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ALC_19</td>\n",
       "      <td>EC:1.1.1.1</td>\n",
       "      <td>2a8b9bb6fab9f06da548a13f0742c437</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.110840</td>\n",
       "      <td>3</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.332521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ALC_19</td>\n",
       "      <td>EC:1.1.1.1</td>\n",
       "      <td>419f4f6586d5b066c1c6361376f1c689</td>\n",
       "      <td>8.5</td>\n",
       "      <td>0.188428</td>\n",
       "      <td>1</td>\n",
       "      <td>8.5</td>\n",
       "      <td>0.188428</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample    function                             taxon  taxon_abun  \\\n",
       "0  ALC_19  EC:1.1.1.1  0746d4ed931f7a6c799e7b1e5f525cbc        29.0   \n",
       "1  ALC_19  EC:1.1.1.1  108039c87f18b4ab5b726b4417cefdca       473.0   \n",
       "2  ALC_19  EC:1.1.1.1  16d4ee39d4468351b3a62dbd64d29ed9        38.0   \n",
       "3  ALC_19  EC:1.1.1.1  2a8b9bb6fab9f06da548a13f0742c437         5.0   \n",
       "4  ALC_19  EC:1.1.1.1  419f4f6586d5b066c1c6361376f1c689         8.5   \n",
       "\n",
       "   taxon_rel_abun  genome_function_count  taxon_function_abun  \\\n",
       "0        0.642873                      1                 29.0   \n",
       "1       10.485480                      2                946.0   \n",
       "2        0.842385                      1                 38.0   \n",
       "3        0.110840                      3                 15.0   \n",
       "4        0.188428                      1                  8.5   \n",
       "\n",
       "   taxon_rel_function_abun  \n",
       "0                 0.642873  \n",
       "1                20.970960  \n",
       "2                 0.842385  \n",
       "3                 0.332521  \n",
       "4                 0.188428  "
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Load stratified metagenome (metagenome by taxon contribution)\n",
    "strat_path = '/home/alexis/UAM/qiime/Mar20/picrust2/EC_metagenome_out/pred_metagenome_contrib.tsv.gz'\n",
    "metagenome_strat = pd.read_csv(strat_path, compression = 'gzip', sep = '\\t')\n",
    "metagenome_strat.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create a ditcionary of dataframes for each sample\n",
    "sample = {}\n",
    "#for each sample in the pre_metagenome\n",
    "for smpl in metagenome_strat.loc[:,'sample'].unique():\n",
    "    #Filter the taxons corresponding to sample i\n",
    "    sample_filter = metagenome_strat.loc[:, 'sample'] == smpl\n",
    "    #Create a dataframe of size = functions x taxons\n",
    "    taxons = metagenome_strat.loc[sample_filter, 'taxon'].unique()\n",
    "    functions = metagenome_strat.loc[sample_filter, 'function'].unique()\n",
    "    sample[smpl] = pd.DataFrame(index = functions, columns  = taxons)\n",
    "    del taxons\n",
    "    del functions\n",
    "    #for every entry in sample i\n",
    "    for idx in metagenome_strat.loc[sample_filter].index:\n",
    "        #query the function abundace\n",
    "        function = metagenome_strat.loc[idx, 'function']\n",
    "        #For every taxon\n",
    "        taxon = metagenome_strat.loc[idx, 'taxon']\n",
    "        #Write it to the corresponding position in the dataframe\n",
    "        sample[smpl].loc[function, taxon] = metagenome_strat.loc[idx, 'taxon_function_abun'].copy()\n",
    "        del function\n",
    "        del taxon\n",
    "    sample[smpl].index.name = 'function'\n",
    "    sample[smpl].reset_index(inplace = True)\n",
    "    sample[smpl].fillna(0, inplace = True)  \n",
    "    #Create directory to save metabolic.net files\n",
    "    #os.mkdir(\"/home/alexis/BiomeNet/BiomeNet/taxon_analysis/metabolic_nets/\"+smpl)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12\n",
      "0\n",
      "1\n",
      "0\n",
      "0\n",
      "0\n",
      "1\n"
     ]
    }
   ],
   "source": [
    "missing_ec = []\n",
    "for smpl in sample.keys():\n",
    "    n = 0\n",
    "    for taxon in sample[smpl].columns[1:]:\n",
    "        #Create a metabolic.net file\n",
    "        metabolic_net = open('/home/alexis/BiomeNet/BiomeNet/taxon_analysis/metabolic_nets/'+smpl+'/'+taxon+'.metabolic.net','w+')\n",
    "        #Filter the reactions with non-zero abundance\n",
    "        nonzero_filter = sample[smpl].loc[:, taxon] > 0\n",
    "        temp_df = sample[smpl].loc[nonzero_filter,['function', taxon]].copy()\n",
    "        #Calculate the centered log-ratio of non-zero reactions\n",
    "        temp_df.loc[:, taxon] = clr(temp_df.loc[:, taxon])\n",
    "        #translate the clr so that all values are positive (this is to be used as the abundance of reaction in metabolic.net files)\n",
    "        temp_df.loc[:, taxon] = temp_df.loc[:, taxon] - 5*temp_df.loc[:, taxon].min()\n",
    "        #For each non-zero abundance reaction\n",
    "        for index in temp_df.index:\n",
    "            #Filter all reactions in the metacyc database annotated with the indexed EC-number in metagenome\n",
    "            EC_filter = EC_reactions.loc[:, 'EC-Number'] == temp_df.loc[index, 'function']\n",
    "            if not EC_reactions.loc[EC_filter].index.empty:\n",
    "                #For all reactions wich matched the EC-number\n",
    "                for rxn in EC_reactions.loc[EC_filter].index:\n",
    "                    #repeat the process n times, where n is the abundace of the indexed reaction\n",
    "                    for i in range(int(round(temp_df.loc[index, taxon]))):\n",
    "                        #Map reactant,reaxionID,product\n",
    "                        for reactant in EC_reactions.loc[rxn, 'Right'].split():\n",
    "                            for product in EC_reactions.loc[rxn, 'Left'].split():\n",
    "                                metabolic_net.write(reactant+','+rxn+','+product+'\\n')\n",
    "                        metabolic_net.write('\\n')\n",
    "            else:\n",
    "                if temp_df.loc[index, 'function'] in missing_ec:\n",
    "                    pass\n",
    "                else:\n",
    "                    n += 1\n",
    "                    missing_ec.append(temp_df.loc[index, 'function'])\n",
    "        del temp_df\n",
    "        metabolic_net.close()\n",
    "    print(n)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2496"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import glob\n",
    "path = r'/home/alexis/BiomeNet/BiomeNet/taxon_analysis/metabolic_nets/CIR_19/' # use your path\n",
    "all_files = glob.glob(os.path.join(path,\"*.metabolic.net\"))\n",
    "df = pd.concat((pd.read_csv(f, header = None) for f in all_files))\n",
    "sample[smpl].loc[nonzero_filter,['function', taxon]]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
