{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from skbio.stats.composition import clr\n",
    "import os\n",
    "import sys\n",
    "os.chdir('/home/alexis/UAM/picrust/Mar20/OTUs_99/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Right</th>\n",
       "      <th>Left</th>\n",
       "      <th>EC-Number</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reaction</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DEOXYNUCLEOTIDE-3-PHOSPHATASE-RXN</th>\n",
       "      <td>Deoxy-Ribonucleosides Pi</td>\n",
       "      <td>Deoxy-Ribonucleoside-3P WATER</td>\n",
       "      <td>EC:3.1.3.34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RXN-20295</th>\n",
       "      <td>CPD-21889 CARBON-DIOXIDE WATER Oxidized-ferred...</td>\n",
       "      <td>CPD-21911 OXYGEN-MOLECULE Reduced-ferredoxins ...</td>\n",
       "      <td>EC:1.14.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RXN-11130</th>\n",
       "      <td>CPD-12101 CO-A</td>\n",
       "      <td>CPD-11571 ACETYL-COA</td>\n",
       "      <td>EC:2.3.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HYPONITRITE-REDUCTASE-RXN</th>\n",
       "      <td>PROTON CPD-1541 NADH</td>\n",
       "      <td>HYDROXYLAMINE NAD</td>\n",
       "      <td>EC:1.7.1.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RXN-19187</th>\n",
       "      <td>CPD-20683 ADP PROTON</td>\n",
       "      <td>CPD-20682 ATP</td>\n",
       "      <td>EC:2.7.1.M16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                               Right  \\\n",
       "Reaction                                                                               \n",
       "DEOXYNUCLEOTIDE-3-PHOSPHATASE-RXN                           Deoxy-Ribonucleosides Pi   \n",
       "RXN-20295                          CPD-21889 CARBON-DIOXIDE WATER Oxidized-ferred...   \n",
       "RXN-11130                                                             CPD-12101 CO-A   \n",
       "HYPONITRITE-REDUCTASE-RXN                                       PROTON CPD-1541 NADH   \n",
       "RXN-19187                                                       CPD-20683 ADP PROTON   \n",
       "\n",
       "                                                                                Left  \\\n",
       "Reaction                                                                               \n",
       "DEOXYNUCLEOTIDE-3-PHOSPHATASE-RXN                      Deoxy-Ribonucleoside-3P WATER   \n",
       "RXN-20295                          CPD-21911 OXYGEN-MOLECULE Reduced-ferredoxins ...   \n",
       "RXN-11130                                                       CPD-11571 ACETYL-COA   \n",
       "HYPONITRITE-REDUCTASE-RXN                                          HYDROXYLAMINE NAD   \n",
       "RXN-19187                                                              CPD-20682 ATP   \n",
       "\n",
       "                                      EC-Number  \n",
       "Reaction                                         \n",
       "DEOXYNUCLEOTIDE-3-PHOSPHATASE-RXN   EC:3.1.3.34  \n",
       "RXN-20295                            EC:1.14.15  \n",
       "RXN-11130                              EC:2.3.1  \n",
       "HYPONITRITE-REDUCTASE-RXN            EC:1.7.1.5  \n",
       "RXN-19187                          EC:2.7.1.M16  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Create Metabolic.net files for analysis in BiomeNet\n",
    "#Metabolic net files have the format reactant,reaxionID,product\n",
    "\n",
    "#Load reactions in the Metacyc Database, annotated with reactants and products\n",
    "EC_reactions = pd.read_csv('/home/alexis/UAM/picrust/default_files/metacyc_rxns_v24.csv', sep = '\\t', index_col = 'Reaction') \n",
    "EC_reactions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>function</th>\n",
       "      <th>ALC_19</th>\n",
       "      <th>ALC_20</th>\n",
       "      <th>CIR_19</th>\n",
       "      <th>My_20</th>\n",
       "      <th>Pty_20</th>\n",
       "      <th>VT_20</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EC:1.1.1.1</td>\n",
       "      <td>10803.0</td>\n",
       "      <td>33055.0</td>\n",
       "      <td>2332.5</td>\n",
       "      <td>30454.25</td>\n",
       "      <td>24851.25</td>\n",
       "      <td>84668.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EC:1.1.1.100</td>\n",
       "      <td>19503.5</td>\n",
       "      <td>76436.0</td>\n",
       "      <td>54364.0</td>\n",
       "      <td>58108.00</td>\n",
       "      <td>84322.00</td>\n",
       "      <td>126711.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EC:1.1.1.103</td>\n",
       "      <td>139.0</td>\n",
       "      <td>5957.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>358.00</td>\n",
       "      <td>5032.00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EC:1.1.1.108</td>\n",
       "      <td>3.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EC:1.1.1.11</td>\n",
       "      <td>36.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>152.0</td>\n",
       "      <td>1747.00</td>\n",
       "      <td>179.00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       function   ALC_19   ALC_20   CIR_19     My_20    Pty_20     VT_20\n",
       "0    EC:1.1.1.1  10803.0  33055.0   2332.5  30454.25  24851.25   84668.8\n",
       "1  EC:1.1.1.100  19503.5  76436.0  54364.0  58108.00  84322.00  126711.2\n",
       "2  EC:1.1.1.103    139.0   5957.0     83.0    358.00   5032.00       0.0\n",
       "3  EC:1.1.1.108      3.5      0.0      6.0      0.00      0.00       0.0\n",
       "4   EC:1.1.1.11     36.0      0.0    152.0   1747.00    179.00       0.0"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metagenome_unstrat = pd.read_csv('EC_metagenome_out/pred_metagenome_unstrat.tsv.gz', compression = 'gzip', sep = '\\t')\n",
    "metagenome_unstrat.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "1\n"
     ]
    }
   ],
   "source": [
    "#Create the files for metabolic divergence analysis between samples\n",
    "#For each sample in the metagenome_file\n",
    "missing_ec = []\n",
    "for sample in metagenome_unstrat.iloc[:,1:].columns:\n",
    "    #Create a metabolic.net file\n",
    "    metabolic_net = open('/home/alexis/BiomeNet/BiomeNet/metabolic_nets/Mar20/'+sample+'.metabolic.net','w')\n",
    "    #Filter the reactions with non-zero abundance\n",
    "    nonzero_filter = metagenome_unstrat.loc[:, sample] > 0\n",
    "    temp_df = metagenome_unstrat.loc[nonzero_filter,['function', sample]]\n",
    "    temp_df.loc[:, sample] = clr(temp_df.loc[:, sample])\n",
    "    temp_df.loc[:, sample] = temp_df.loc[:, sample] - 5*temp_df.loc[:, sample].min()\n",
    "    #For each non-zero abundance reaction\n",
    "    n = 0\n",
    "    for index in temp_df.index:\n",
    "        #Filter all reactions in the metacyc database annotated with indexed EC-number in metagenome\n",
    "        EC_filter = EC_reactions.loc[:, 'EC-Number'] == temp_df.loc[index, 'function']\n",
    "        if not EC_reactions.loc[EC_filter].index.empty:\n",
    "            #For all reactions wich matched the EC-number\n",
    "            for rxn in EC_reactions.loc[EC_filter].index:\n",
    "                #repeat the process n times, where n is the abundace of the indexed reaction\n",
    "                for i in range(int(round(temp_df.loc[index, sample]))):\n",
    "                    #Map reactant,reaxionID,product\n",
    "                    for reactant in EC_reactions.loc[rxn, 'Right'].split():\n",
    "                        for product in EC_reactions.loc[rxn, 'Left'].split():\n",
    "                            metabolic_net.write(reactant+','+rxn+','+product+'\\n')\n",
    "                    metabolic_net.write('\\n')\n",
    "        else:\n",
    "            if temp_df.loc[index, 'function'] in missing_ec:\n",
    "                pass\n",
    "            else:\n",
    "                n += 1\n",
    "                missing_ec.append(temp_df.loc[index, 'function'])\n",
    "    print(n)\n",
    "    del temp_df\n",
    "    metabolic_net.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>function</th>\n",
       "      <th>taxon</th>\n",
       "      <th>taxon_abun</th>\n",
       "      <th>taxon_rel_abun</th>\n",
       "      <th>genome_function_count</th>\n",
       "      <th>taxon_function_abun</th>\n",
       "      <th>taxon_rel_function_abun</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ALC_19</td>\n",
       "      <td>EC:1.1.1.1</td>\n",
       "      <td>006e3ba76f4a4306769c277678e3bad8</td>\n",
       "      <td>72.00</td>\n",
       "      <td>1.125294</td>\n",
       "      <td>2</td>\n",
       "      <td>144.00</td>\n",
       "      <td>2.250587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ALC_19</td>\n",
       "      <td>EC:1.1.1.1</td>\n",
       "      <td>088408debd5cf6b712322ae0f6ab61c5</td>\n",
       "      <td>32.50</td>\n",
       "      <td>0.507945</td>\n",
       "      <td>1</td>\n",
       "      <td>32.50</td>\n",
       "      <td>0.507945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ALC_19</td>\n",
       "      <td>EC:1.1.1.1</td>\n",
       "      <td>08935de8e561e37b02f87dbf466da7a7</td>\n",
       "      <td>51.50</td>\n",
       "      <td>0.804898</td>\n",
       "      <td>2</td>\n",
       "      <td>103.00</td>\n",
       "      <td>1.609795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ALC_19</td>\n",
       "      <td>EC:1.1.1.1</td>\n",
       "      <td>0b345a10cf9aefee3326b1300b680a15</td>\n",
       "      <td>48.33</td>\n",
       "      <td>0.755353</td>\n",
       "      <td>1</td>\n",
       "      <td>48.33</td>\n",
       "      <td>0.755353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ALC_19</td>\n",
       "      <td>EC:1.1.1.1</td>\n",
       "      <td>0b9c0802af21fbaf35a7e303cf5a2d5d</td>\n",
       "      <td>26.50</td>\n",
       "      <td>0.414171</td>\n",
       "      <td>1</td>\n",
       "      <td>26.50</td>\n",
       "      <td>0.414171</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample    function                             taxon  taxon_abun  \\\n",
       "0  ALC_19  EC:1.1.1.1  006e3ba76f4a4306769c277678e3bad8       72.00   \n",
       "1  ALC_19  EC:1.1.1.1  088408debd5cf6b712322ae0f6ab61c5       32.50   \n",
       "2  ALC_19  EC:1.1.1.1  08935de8e561e37b02f87dbf466da7a7       51.50   \n",
       "3  ALC_19  EC:1.1.1.1  0b345a10cf9aefee3326b1300b680a15       48.33   \n",
       "4  ALC_19  EC:1.1.1.1  0b9c0802af21fbaf35a7e303cf5a2d5d       26.50   \n",
       "\n",
       "   taxon_rel_abun  genome_function_count  taxon_function_abun  \\\n",
       "0        1.125294                      2               144.00   \n",
       "1        0.507945                      1                32.50   \n",
       "2        0.804898                      2               103.00   \n",
       "3        0.755353                      1                48.33   \n",
       "4        0.414171                      1                26.50   \n",
       "\n",
       "   taxon_rel_function_abun  \n",
       "0                 2.250587  \n",
       "1                 0.507945  \n",
       "2                 1.609795  \n",
       "3                 0.755353  \n",
       "4                 0.414171  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Load stratified metagenome (metagenome by taxon contribution)\n",
    "strat_path = 'EC_metagenome_out/pred_metagenome_contrib.tsv.gz'\n",
    "metagenome_strat = pd.read_csv(strat_path, compression = 'gzip', sep = '\\t')\n",
    "metagenome_strat.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create a ditcionary of dataframes for each sample\n",
    "sample = {}\n",
    "#for each sample in the pre_metagenome\n",
    "for smpl in metagenome_strat.loc[:,'sample'].unique():\n",
    "    #Filter the taxons corresponding to sample i\n",
    "    sample_filter = metagenome_strat.loc[:, 'sample'] == smpl\n",
    "    #Create a dataframe of size = functions x taxons\n",
    "    taxons = metagenome_strat.loc[sample_filter, 'taxon'].unique()\n",
    "    functions = metagenome_strat.loc[sample_filter, 'function'].unique()\n",
    "    sample[smpl] = pd.DataFrame(index = functions, columns  = taxons)\n",
    "    del taxons\n",
    "    del functions\n",
    "    #for every entry in sample i\n",
    "    for idx in metagenome_strat.loc[sample_filter].index:\n",
    "        #query the function abundace\n",
    "        function = metagenome_strat.loc[idx, 'function']\n",
    "        #For every taxon\n",
    "        taxon = metagenome_strat.loc[idx, 'taxon']\n",
    "        #Write it to the corresponding position in the dataframe\n",
    "        sample[smpl].loc[function, taxon] = metagenome_strat.loc[idx, 'taxon_function_abun'].copy()\n",
    "        del function\n",
    "        del taxon\n",
    "    sample[smpl].index.name = 'function'\n",
    "    sample[smpl].reset_index(inplace = True)\n",
    "    sample[smpl].fillna(0, inplace = True)  \n",
    "    #Create directory to save metabolic.net files\n",
    "    os.mkdir(\"/home/alexis/BiomeNet/BiomeNet/taxon_analysis/metabolic_nets/OTUs_99/\"+smpl)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ALC_19 13\n",
      "ALC_20 0\n",
      "CIR_19 0\n",
      "My_20 0\n",
      "Pty_20 1\n",
      "V3_20 1\n",
      "VT_20 1\n"
     ]
    }
   ],
   "source": [
    "missing_ec = []\n",
    "for smpl in sample.keys():\n",
    "    n = 0\n",
    "    for taxon in sample[smpl].columns[1:]:\n",
    "        #Create a metabolic.net file\n",
    "        metabolic_net = open('/home/alexis/BiomeNet/BiomeNet/taxon_analysis/metabolic_nets/OTUs_99/'+smpl+'/'+taxon+'.metabolic.net','w+')\n",
    "        #Filter the reactions with non-zero abundance\n",
    "        nonzero_filter = sample[smpl].loc[:, taxon] > 0\n",
    "        temp_df = sample[smpl].loc[nonzero_filter,['function', taxon]].copy()\n",
    "        #Calculate the centered log-ratio of non-zero reactions\n",
    "        temp_df.loc[:, taxon] = clr(temp_df.loc[:, taxon])\n",
    "        #translate the clr so that all values are positive (this is to be used as the abundance of reaction in metabolic.net files)\n",
    "        temp_df.loc[:, taxon] = temp_df.loc[:, taxon] - 5*temp_df.loc[:, taxon].min()\n",
    "        #For each non-zero abundance reaction\n",
    "        for index in temp_df.index:\n",
    "            #Filter all reactions in the metacyc database annotated with the indexed EC-number in metagenome\n",
    "            EC_filter = EC_reactions.loc[:, 'EC-Number'] == temp_df.loc[index, 'function']\n",
    "            if not EC_reactions.loc[EC_filter].index.empty:\n",
    "                #For all reactions wich matched the EC-number\n",
    "                for rxn in EC_reactions.loc[EC_filter].index:\n",
    "                    #repeat the process n times, where n is the abundace of the indexed reaction\n",
    "                    for i in range(int(round(temp_df.loc[index, taxon]))):\n",
    "                        #Map reactant,reaxionID,product\n",
    "                        for reactant in EC_reactions.loc[rxn, 'Right'].split():\n",
    "                            for product in EC_reactions.loc[rxn, 'Left'].split():\n",
    "                                metabolic_net.write(reactant+','+rxn+','+product+'\\n')\n",
    "                        metabolic_net.write('\\n')\n",
    "            else:\n",
    "                if temp_df.loc[index, 'function'] in missing_ec:\n",
    "                    pass\n",
    "                else:\n",
    "                    n += 1\n",
    "                    missing_ec.append(temp_df.loc[index, 'function'])\n",
    "        del temp_df\n",
    "        metabolic_net.close()\n",
    "    print(smpl, n)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2496"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import glob\n",
    "path = r'/home/alexis/BiomeNet/BiomeNet/taxon_analysis/metabolic_nets/CIR_19/' # use your path\n",
    "all_files = glob.glob(os.path.join(path,\"*.metabolic.net\"))\n",
    "df = pd.concat((pd.read_csv(f, header = None) for f in all_files))\n",
    "len(df.loc[:,1].unique())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
